app/javascript/packages/document-capture/components/submission.jsx

```diff
@@ -6,6 +6,40 @@ import SubmissionComplete from './submission-complete';
 import SubmissionInterstitial from './submission-interstitial';
 import CallbackOnMount from './callback-on-mount';
 
+/**
+ * Given an object where some values may be promises, returns a promise which resolves to an object
+ * with the same keys and all resolved promise values.
+ *
+ * @template {Record<string, any>} T
+ *
+ * @param {T} object
+ *
+ * @return {Promise<Record<keyof T, any>>} Object with all values resolved.
+ */
+export async function resolveObjectValues(object) {
```

Comment:
IMO this is not necessarily something we need to write? `Promise.all` can do the same thing with a deconstruct, right? And `Promise.all` already handles immediate values:

```js
Promise.all([1, 2]).then(([a, b]) => { console.log(a); console.log(b) })
// 1
// 2
```

Reply:
> IMO this is not necessarily something we need to write? `Promise.all` can do the same thing with a deconstruct, right? And `Promise.all` already handles immediate values:

Not sure I follow exactly, but yes, if all we'd cared about is awaiting all values to be resolved, we could do something as simple as...

```js
Promise.all(Object.values(object));
```

But the thinking here is not just to await the resolution, but then to use the resolved value as that which is submitted in the payload.

It does bring up a good point, which is that we don't actually want anything to be submitted for these background uploads, so...

- [The promise](https://github.com/18F/identity-idp/blob/ad50f6f2af36c1e1fa3561139182831d7faa4f0a/app/javascript/packages/document-capture/higher-order/with-background-encrypted-upload.jsx#L16) should resolve to `undefined`, and [submission should omit `undefined` values](https://github.com/18F/identity-idp/blob/ad50f6f2af36c1e1fa3561139182831d7faa4f0a/app/javascript/packages/document-capture/services/upload.js#L24).
- Or, we consider a different way to "hold" the submission on these pending uploads. I think it's fair to say it's a bit odd to use the form data object this way for values we have no intention of including in the payload. At the same time, there's a huge convenience in using it to allow the values to be maintained between steps, and to leverage existing error handling built-in to the `Submission` component. And I think there certainly _could_ be cases where we'd actually want to use a resolved promise value in a form submission. For example, I'd thought it'd be a nice way to avoid where we currently [block the user from continuing after capture and before crop](https://github.com/18F/identity-idp/blob/3ecdadc7e16ed40a8e6471e59c3fd41183e50317/app/javascript/packages/document-capture/components/acuant-capture-canvas.jsx#L120-L121) in the Acuant capture flow (Acuant even recommends to ["show a loading screen until onCropped is called"](https://github.com/Acuant/JavascriptWebSDKV11/tree/master/SimpleHTMLApp#acuantcameraui))

There might still be a more expressive way to implement the function, e.g. using `Array#reduce`.

Reply:
> In any case, there might still be a more expressive way to implement the function, e.g. using `Array#reduce`.

This works, but I don't feel that it's especially easy to interpret:

```js
export async function resolveObjectValues(object) {
  return Object.entries(object).reduce(async (current, [key, value]) => {
    const next = await current;
    next[key] = await value;
    return next;
  }, {});
}
```

Reply:
> * [The promise](https://github.com/18F/identity-idp/blob/ad50f6f2af36c1e1fa3561139182831d7faa4f0a/app/javascript/packages/document-capture/higher-order/with-background-encrypted-upload.jsx#L16) should resolve to `undefined`,

Updated in a3665668b.

> and [submission should omit `undefined` values](https://github.com/18F/identity-idp/blob/ad50f6f2af36c1e1fa3561139182831d7faa4f0a/app/javascript/packages/document-capture/services/upload.js#L24).

Updated in 416b5d9cc.